### PR TITLE
Update OpenBLAS commit 

### PR DESCRIPTION
**Pull request by aditew01**
Motivation: Update OpenBLAS and change build script to enable SBGEMM kernels . Update pytorch `jammy` builds for aarch64 to use `install_openblas.sh` instead of `conda_install` 

Link to full [TorchInductor Performance Dashboard AArch64](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2006%20Jun%202025%2009%3A46%3A35%20GMT&stopTime=Fri%2C%2013%20Jun%202025%2009%3A46%3A35%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cpu%20(aarch64)&lBranch=adi/update_openblas&lCommit=0218b65bcf61971c1861cfe8bc586168b73aeb5f&rBranch=main&rCommit=9d59b516e9b3026948918e3ff8c2ef55a33d13ad)

1. This shows a promising speedup across most of the HF models in benchmark, specifically giving a significant boost to SDPA layers. 
2. Overall torch-bench pass-rate (cpp_wrapper mode) increased `[87%, 65/75 → 96%, 72/75]`  

Screenshot 2025-06-20 at 17 05 15

cc @malfet @snadampal @milpuz01 @nikhil-arm @fadara01

### CODE REVIEW

**Comment by albanD** [comment]
Can you give details on why this would change the pass rate?
Also we usually only pin to released versions.

**Comment by aditew01** [comment]
> Can you give details on why this would change the pass rate? 
That's something which I was wondering as well, turns out some models were getting timeout. I assume it's because of enough speedup for these models to pass. `tts_angular` seem to just pass. 

> Also we usually only pin to released versions.

I see, [OpenBLAS 0.30.0](https://github.com/OpenMathLib/OpenBLAS/milestone/49) seems close to release. Do you think it's a good idea to wait? 

Quick question about the conda install for OpenBLAS—does using the install script I proposed look good in terms of how it’s structured in the code. ?

**Comment by albanD** [comment]
I'll let @seemethere comment on the CI script organization :D

**Review by malfet — APPROVED** [review]
This only affects CI, so SGTM

**Comment by aditew01** [comment]
To update on the failures, we run into segfault issue when `train=True`.  We'll be debugging this further. tag: @taoye9

**Comment by aditew01** [comment]
Update regarding the segfaults and SDPA failing tests (train=True):
 
We observed intermittent segmentation faults after integrating the SBGEMM kernels for the backward pass, specifically occurring at this location: [CPUBlas.cpp](https://github.com/pytorch/pytorch/blob/eecaa0bbc6a39c89366f9ec97797479c27f6d760/aten/src/ATen/native/CPUBlas.cpp#L364). Additionally, memory corruption has been identified after the conversion path and transpose APIs here: [Attention.cpp](https://github.com/pytorch/pytorch/blob/eecaa0bbc6a39c89366f9ec97797479c27f6d760/aten/src/ATen/native/transformers/attention.cpp#L972).

It is currently unclear whether these segfaults stem from memory corruption issues within OpenBLAS or conversion. We're  investigating this.

Due to the significant inference speed improvements achieved for SDPA on AArch64, I’ve temporarily disabled this problematic route through a patch. The CI tests are now passing. 
[Latest HUD benchmark](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2006%20Jun%202025%2009%3A46%3A35%20GMT&stopTime=Fri%2C%2013%20Jun%202025%2009%3A46%3A35%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cpu%20(aarch64)&lBranch=adi/update_openblas&lCommit=0218b65bcf61971c1861cfe8bc586168b73aeb5f&rBranch=main&rCommit=9d59b516e9b3026948918e3ff8c2ef55a33d13ad) results on this branch. 
Is there any other CI pipeline we would like to test this on?
If all the tests looks good, can we please merge this?

A follow-up fix to permanently address the underlying issue will soon follow.



cc: @milpuz01

**Comment by aditew01** [comment]
The segfault which we were seeing earlier seems to be resolved now. Thanks @Ryo-not-rio . Although with the latest runs, seeing some extra failures which seems unrelated - related open issue: https://github.com/pytorch/pytorch/issues/155607

cc: @milpuz01 @malfet

File: aten/src/ATen/native/CPUBlas.cpp [hunk 1]
```diff
@@ -368,8 +368,12 @@ void gemm(
               b, &ldb_,
               &beta_,
               float_v.data(), &ldc_);
-      for (auto cv: float_v) {
-        *(c++) = c10::convert<at::BFloat16>(cv);
+
+      for (const auto j : c10::irange(n)) {
+        for (const auto i : c10::irange(m)) {
+          auto offset = j * ldc + i;
+          c[offset] = c10::convert<c10::BFloat16>(float_v[j * m + i]);
```
Comment (taoye9) [inline aten/src/ATen/native/CPUBlas.cpp]: should it be `float_v[j * ldc_ + i]`  instead of `float_v[j * m + i]` given sbgemm leading dim is ldc?

File: aten/src/ATen/native/CPUBlas.cpp [hunk 2]
```diff
@@ -368,8 +368,12 @@ void gemm(
               b, &ldb_,
               &beta_,
               float_v.data(), &ldc_);
-      for (auto cv: float_v) {
-        *(c++) = c10::convert<at::BFloat16>(cv);
+
+      for (const auto j : c10::irange(n)) {
+        for (const auto i : c10::irange(m)) {
+          auto offset = j * ldc + i;
```
Comment (taoye9) [inline aten/src/ATen/native/CPUBlas.cpp]: it's better to move j * ldc out of inner loop to reduce unnecessary multiplication.

**Comment by taoye9** [comment]
I think the cpublas::gemm has a bug in it which causes the segfault. I've opened an issue to explain the causes and potential fixup. https://github.com/pytorch/pytorch/issues/156022. it seems a legacy pytorch bug, not related to openblas.
@aditew01 @Ryo-not-rio

File: aten/src/ATen/native/CPUBlas.cpp [hunk 3]
```diff
@@ -368,8 +368,12 @@ void gemm(
               b, &ldb_,
               &beta_,
               float_v.data(), &ldc_);
-      for (auto cv: float_v) {
-        *(c++) = c10::convert<at::BFloat16>(cv);
+
+      for (const auto j : c10::irange(n)) {
```
Comment (taoye9) [inline aten/src/ATen/native/CPUBlas.cpp]: The output grads in cpu sdpa unit test are always null: https://github.com/pytorch/pytorch/blob/main/test/test_transformers.py#L2138-L2141 

Therefore, it can not detect wether backward's numeric results are correct or not. Need to find a way to fix this ci bug.

**Comment by aditew01** [comment]
@malfet do you think we can get a review for this? 
It would be great to merge this before `2.8 release` branch-cut.   

Note: With the latest runs, seeing some extra failures which seems unrelated - related open issue: https://github.com/pytorch/pytorch/issues/155607

File: aten/src/ATen/native/CPUBlas.cpp (hunk 3 above)
Comment (aditew01) [reply aten/src/ATen/native/CPUBlas.cpp]: tracking this as a part: https://github.com/pytorch/pytorch/issues/156022

**Comment by malfet** [comment]
> @malfet do you think we can get a review for this? It would be great to merge this before `2.8 release` branch-cut.

I see there are some failures, that you are still fixing. Don't you think then it's a bit premature to ask for a review? 
Also, do you mind to re-run the benchmark and update the numbers (ones you are posting are from Apr 17th)

**Comment by aditew01** [comment]
>  The CI tests are now passing. [Latest HUD benchmark](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2006%20Jun%202025%2009%3A46%3A35%20GMT&stopTime=Fri%2C%2013%20Jun%202025%2009%3A46%3A35%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cpu%20(aarch64)&lBranch=adi/update_openblas&lCommit=0218b65bcf61971c1861cfe8bc586168b73aeb5f&rBranch=main&rCommit=9d59b516e9b3026948918e3ff8c2ef55a33d13ad) results on this branch. 

@malfet I commented a more recent run of the branch (13/06). Looks like the fix for the segfault which was pushed earlier have been reverted in a most recent cleanup. I'll push a fix for this. The issue had been identified and we've raised a general issue regarding this as well. Issue reference: https://github.com/pytorch/pytorch/issues/156022
I'll revert the bad change. I'll re-request a review after the CI passes. Thanks.

**Comment by aditew01** [comment]
@malfet the CI runs LGTM. Failures are unrelated to the changes. It's ready for a review.
I've updated the description with a more recent dashboard runs. Please can I get a fresh review,

**Review by snadampal — APPROVED** [review]

**Review by fadara01 — APPROVED** [review]
LGTM!
@malfet could you please have another look at this?

**Review by malfet — APPROVED** [review]
